database.cpp

```diff
@@ -4371,38 +4378,34 @@ static int sqliteLoadAllGatewaysCallback(void *user, int ncols, char **colval ,
 
 /*! Determines a unused id for a sensor.
  */
-int DeRestPluginPrivate::getFreeSensorId()
+int getFreeSensorId()
 {
-    int rc;
-    bool ok;
-    char *errmsg = 0;
+    DeRestPluginPrivate *plugin = DeRestPluginPrivate::instance();
 
-    DBG_Assert(db != 0);
+    DBG_Assert(plugin && plugin->dbIsOpen());
 
-    if (!db)
+    if (!plugin || !plugin->dbIsOpen())
     {
-        return 1;
+        DBG_Printf(DBG_ERROR, "DB getFreeSensorId() called with no valid db pointer\n");
+        return 1; // TODO, this is an error we should handle this. 1 is misleading
     }
 
-    sensorIds.clear();
+    std::vector<int> sensorIds(plugin->sensors.size());
 
-    { // append all ids from nodes known at runtime
-        std::vector<Sensor>::const_iterator i = sensors.begin();
-        std::vector<Sensor>::const_iterator end = sensors.end();
-        for (;i != end; ++i)
-        {
-            sensorIds.push_back(i->id().toUInt());
-        }
-    }
+    // collect all ids from nodes known at runtime
+    std::transform (plugin->sensors.cbegin(), plugin->sensors.cend(), sensorIds.begin(),
+        [](const Sensor&s) { return s.id().toInt(); }
+    );
 
-    // add all ids references in rules
-    for (const Rule &r : rules)
+    // add all ids referenced in rules of sensors which don't exist anymore -> to not consider these
+    for (const Rule &r : plugin->rules)
     {
         for (const RuleCondition &c : r.conditions())
         {
-            if (c.address().startsWith(QLatin1String("/sensors/")))
+            if (c.resource() == RSensors)
             {
-                uint id = c.id().toUInt(&ok);
+                bool ok;
+                const int id = c.id().toInt(&ok);

```

Comment:
You might wanna use `sid` instead of `id` here and in the following as well for consistency

Reply:
check :)